cfgmgr/buffermgrdyn.cpp

```diff
@@ -1173,10 +1351,46 @@ task_process_status BufferMgrDynamic::handleBufferPoolTable(KeyOpFieldsValuesTup
             {
                 bufferPool.ingress = (value == buffer_value_ingress);
             }
-            fvVector.emplace_back(FieldValueTuple(field, value));
+            fvVector.emplace_back(field, value);
             SWSS_LOG_INFO("Inserting BUFFER_POOL table field %s value %s", field.c_str(), value.c_str());
         }
-        if (!bufferPool.dynamic_size)
+
+        bool dontUpdatePoolToDb = bufferPool.dynamic_size;
+        if (pool == INGRESS_LOSSLESS_PG_POOL_NAME)
+        {
+            /*
+             * "dontUpdatPoolToDb" is calculated for ingress_lossless_pool according to following rules:
+             * Don't update | pool size | SHP enabled by size | SHP enabled by over subscribe ratio
+             * True         | Dynamic   | Any                 | Any
+             * False        | Static    | True                | Any
+             * True         | Static    | False               | True
+             * False        | Static    | False               | False
+             */
+            bool willSHPBeEnabledBySize = isNonZero(newSHPSize);
+            if (newSHPSize != m_configuredSharedHeadroomPoolSize)
+            {
+                bool isSHPEnabledBySize = isNonZero(m_configuredSharedHeadroomPoolSize);
+
+                m_configuredSharedHeadroomPoolSize = newSHPSize;
+                refreshSharedHeadroomPool(false, isSHPEnabledBySize != willSHPBeEnabledBySize);
+            }
+            else if (!newSHPSize.empty())
+            {
+                SWSS_LOG_INFO("Shared headroom pool size updated without change (new %s vs current %s), skipped", newSHPSize.c_str(), m_configuredSharedHeadroomPoolSize.c_str());
+            }
+
+            if (!willSHPBeEnabledBySize)
+            {
+                // Don't need to program APPL_DB if shared headroom pool is enabled
+                dontUpdatePoolToDb |= isNonZero(m_overSubscribeRatio);
+            }
+        }
+        else if (!isNonZero(newSHPSize))
+        {
+            SWSS_LOG_ERROR("Field xoff is supported for %s only, but got for %s, ignored", INGRESS_LOSSLESS_PG_POOL_NAME, pool.c_str());
+        }
+
+        if (!dontUpdatePoolToDb)
```

Comment:
The `xoff` will be ignored if the pool is **not** `ingress_lossless_pool` regardless of whether its size is static or not. This is enforced by the check in L1388. (Thank you for pointing out the typo in the condition :)
Meanwhile, we don't need to add check in line 456 because the callers guarantee `xoff` only occurs in `ingress_lossless_pool`. In both `recalculateSharedBufferPool` and `refreshSharedHeadroomPool`,  `xoff` is updated for `ingress_lossless_pool` only.